subsys/bluetooth/adv_prov/providers/fast_pair.c

```diff
@@ -41,37 +60,97 @@ int bt_le_adv_prov_fast_pair_set_battery_mode(enum bt_fast_pair_adv_battery_mode
+static enum bt_fast_pair_adv_mode get_adv_mode(bool pairing_mode, bool show_ui)
 {
+	enum bt_fast_pair_adv_mode adv_mode;
+
+	if (pairing_mode) {
+		adv_mode = BT_FAST_PAIR_ADV_MODE_DISCOVERABLE;
+	} else {
+		if (show_ui) {
+			adv_mode = BT_FAST_PAIR_ADV_MODE_NOT_DISCOVERABLE_SHOW_UI_IND;
+		} else {
+			adv_mode = BT_FAST_PAIR_ADV_MODE_NOT_DISCOVERABLE_HIDE_UI_IND;
+		}
+	}
+
+	return adv_mode;
+}
+
+static int generate_payload(enum bt_fast_pair_adv_mode adv_mode,
+			    enum bt_fast_pair_adv_battery_mode battery_mode)
+{
+	int err;
 	static uint8_t buf[ADV_DATA_BUF_SIZE];
 	struct bt_fast_pair_adv_config adv_config;
 
+	__ASSERT_NO_MSG(bt_fast_pair_adv_data_size(adv_config) <= sizeof(buf));
+
+	adv_config.adv_mode = adv_mode;
+	adv_config.adv_battery_mode = battery_mode;
+
+	err = bt_fast_pair_adv_data_fill(&gen_ad, buf, sizeof(buf), adv_config);
+	if (!err) {
+		memcpy(&gen_adv_config, &adv_config, sizeof(gen_adv_config));
+	} else {
+		generated_adv_payload_invalidate();
+	}
+
+	return err;
+}
+
+static int get_data(struct bt_data *ad, const struct bt_le_adv_prov_adv_state *state,
+		    struct bt_le_adv_prov_feedback *fb)
+{
+	int err = 0;
+	bool generate_new_payload = false;
+	enum bt_fast_pair_adv_mode adv_mode;
+
 	if (!enabled) {
 		return -ENOENT;
 	}
 
-	if (state->pairing_mode) {
-		adv_config.adv_mode = BT_FAST_PAIR_ADV_MODE_DISCOVERABLE;
-	} else {
-		if (show_ui_pairing) {
-			adv_config.adv_mode = BT_FAST_PAIR_ADV_MODE_NOT_DISCOVERABLE_SHOW_UI_IND;
-		} else {
-			adv_config.adv_mode = BT_FAST_PAIR_ADV_MODE_NOT_DISCOVERABLE_HIDE_UI_IND;
+	if (IS_ENABLED(CONFIG_BT_ADV_PROV_FAST_PAIR_STOP_DISCOVERABLE_ON_RPA_ROTATION) &&
+	    (gen_adv_config.adv_mode == BT_FAST_PAIR_ADV_MODE_DISCOVERABLE)) {
+		static bool adv_session_rpa_rotated;
+
+		if (state->new_adv_session) {
+			adv_session_rpa_rotated = false;
+		} else if (state->rpa_rotated || adv_session_rpa_rotated) {
+			adv_session_rpa_rotated = true;
+
+			LOG_WRN("RPA rotated during Fast Pair discoverable advertising session");
+			LOG_WRN("Removed Fast Pair advertising payload");
+			return -ENOENT;
 		}
 	}
 
-	adv_config.adv_battery_mode = adv_battery_mode;
+	adv_mode = get_adv_mode(state->pairing_mode, show_ui_pairing);
+	if (adv_mode != gen_adv_config.adv_mode) {
+		generate_new_payload = true;
+	} else if (
+		(gen_adv_config.adv_mode == BT_FAST_PAIR_ADV_MODE_NOT_DISCOVERABLE_SHOW_UI_IND) ||
+		(gen_adv_config.adv_mode == BT_FAST_PAIR_ADV_MODE_NOT_DISCOVERABLE_HIDE_UI_IND)) {
+		if (state->rpa_rotated || (gen_adv_config.adv_battery_mode != adv_battery_mode)) {
+			generate_new_payload = true;
+		}
+	}
 
-	if (IS_ENABLED(CONFIG_BT_ADV_PROV_FAST_PAIR_AUTO_SET_PAIRING_MODE)) {
-		/* Set Fast Pair pairing mode to match Fast Pair advertising mode. */
-		bt_fast_pair_set_pairing_mode(adv_config.adv_mode ==
-					      BT_FAST_PAIR_ADV_MODE_DISCOVERABLE);
+	if (generate_new_payload) {
+		err = generate_payload(adv_mode, adv_battery_mode);
```

Comment:
Couldn't we always generate new payload? We are relying here on the implementation of `bt_fast_pair_adv_data_fill` function and on the fact that in some cases the advertising payload doesn't change. I don't like that dependency.

Reply:
Maybe we could also allow to force generating new data (e.g. introduce a new API for the user). User may need to re-generate data e.g. when battery value was updated or when new Account Key was written.

Reply:
We could discuss this if needed.